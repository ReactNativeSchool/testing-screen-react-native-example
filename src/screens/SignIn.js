import React, { useState } from "react";
import { StyleSheet, Text, KeyboardAvoidingView } from "react-native";

import { Input, Button, ErrorText } from "../components/Form";

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff",
    justifyContent: "center",
    paddingHorizontal: 40,
  },
  headerText: {
    color: "#353031",
    fontWeight: "bold",
    fontSize: 34,
    marginBottom: 10,
  },
});

const useLoginFormState = ({ navigation }) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [submit, setSubmit] = useState(false);

  let isUsernameValid = false;
  let isPasswordValid = false;

  if (username === "example") {
    isUsernameValid = true;
  }

  if (password === "asdf") {
    isPasswordValid = true;
  }

  return {
    username: {
      value: username,
      set: setUsername,
      valid: isUsernameValid,
    },
    password: {
      value: password,
      set: setPassword,
      valid: isPasswordValid,
    },
    submit: {
      value: submit,
      set: () => {
        setSubmit(true);

        if (isUsernameValid && isPasswordValid) {
          fetch("https://jsonplaceholder.typicode.com/users", {
            method: "POST",
            body: JSON.stringify({
              username,
              password,
            }),
          })
            .then((response) => response.json())
            .then(() => {
              navigation.push("App");
            })
            .catch((error) => {
              console.log("error", error);
            });
        }
      },
    },
  };
};

export default ({ navigation }) => {
  const { username, password, submit } = useLoginFormState({ navigation });

  let usernameErrorMsg;
  let passwordErrorMsg;

  if (submit.value && !username.valid) {
    usernameErrorMsg = "Invalid username.";
  }

  if (submit.value && !password.valid) {
    passwordErrorMsg = "Invalid password.";
  }

  return (
    <KeyboardAvoidingView style={styles.container} behavior="position">
      <Text style={styles.headerText}>Login</Text>
      <Input
        label="Username"
        placeholder="example"
        onChangeText={username.set}
        error={usernameErrorMsg}
        testID="SignIn.usernameInput"
      />
      <Input
        label="Password"
        placeholder="***"
        secureTextEntry
        onChangeText={password.set}
        error={passwordErrorMsg}
        testID="SignIn.passwordInput"
      />
      <ErrorText messages={[usernameErrorMsg, passwordErrorMsg]} />
      <Button testID="SignIn.Button" text="Login" onPress={submit.set} />
    </KeyboardAvoidingView>
  );
};
