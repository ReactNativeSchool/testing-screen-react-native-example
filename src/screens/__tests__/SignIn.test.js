import React from "react";
import { render, fireEvent, act } from "@testing-library/react-native";

import SignIn from "../SignIn";

const flushMicrotasksQueue = () =>
  new Promise((resolve) => setImmediate(resolve));

it("renders default elements", () => {
  const { getAllByText, getByPlaceholderText } = render(<SignIn />);

  expect(getAllByText("Login").length).toBe(2);
  getByPlaceholderText("example");
  getByPlaceholderText("***");
});

it("shows invalid input messages", () => {
  const { getByTestId, getByText } = render(<SignIn />);

  fireEvent.press(getByTestId("SignIn.Button"));

  getByText("Invalid username.");
  getByText("Invalid password.");
});

it("shows invalid user name error message", () => {
  const { getByTestId, getByText, queryAllByText } = render(<SignIn />);

  fireEvent.changeText(getByTestId("SignIn.passwordInput"), "asdf");

  fireEvent.press(getByTestId("SignIn.Button"));

  getByText("Invalid username.");
  expect(queryAllByText("Invalid password.").length).toBe(0);

  fireEvent.changeText(getByTestId("SignIn.usernameInput"), "invalid input");

  getByText("Invalid username.");
  expect(queryAllByText("Invalid password.").length).toBe(0);
});

it("shows invalid password error message", () => {
  const { getByTestId, getByText, queryAllByText } = render(<SignIn />);

  fireEvent.changeText(getByTestId("SignIn.usernameInput"), "example");

  fireEvent.press(getByTestId("SignIn.Button"));

  getByText("Invalid password.");
  expect(queryAllByText("Invalid username.").length).toBe(0);

  fireEvent.changeText(getByTestId("SignIn.passwordInput"), "invalid input");

  getByText("Invalid password.");
  expect(queryAllByText("Invalid username.").length).toBe(0);
});

it("handles valid input submission", async () => {
  fetch.mockResponseOnce(JSON.stringify({ passes: true }));

  const pushMock = jest.fn();
  const { getByTestId } = render(<SignIn navigation={{ push: pushMock }} />);

  fireEvent.changeText(getByTestId("SignIn.usernameInput"), "example");
  fireEvent.changeText(getByTestId("SignIn.passwordInput"), "asdf");
  fireEvent.press(getByTestId("SignIn.Button"));

  expect(fetch.mock.calls).toMatchSnapshot();
  await act(flushMicrotasksQueue);

  expect(pushMock).toBeCalledWith("App");
});
