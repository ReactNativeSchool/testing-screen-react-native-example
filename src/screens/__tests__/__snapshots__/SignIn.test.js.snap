// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles valid input submission 1`] = `
Array [
  Array [
    "https://jsonplaceholder.typicode.com/users",
    Object {
      "body": "{\\"username\\":\\"example\\",\\"password\\":\\"asdf\\"}",
      "method": "POST",
    },
  ],
]
`;
